src/equations/shallow_water_2d.jl

```diff
@@ -0,0 +1,619 @@
+# By default, Julia/LLVM does not use fused multiply-add operations (FMAs).
+# Since these FMAs can increase the performance of many numerical algorithms,
+# we need to opt-in explicitly.
+# See https://ranocha.de/blog/Optimizing_EC_Trixi for further details.
+@muladd begin
+
+
+@doc raw"""
+    ShallowWaterEquations2D(gravity)
+
+Shallow water equations (SWE) in two space dimensions. The equations are given by
+```math
+\begin{aligned}
+  \frac{\partial h}{\partial t} + \frac{\partial}{\partial x}(h v_1)
+    + \frac{\partial}{\partial y}(h v_2) &= 0 \\
+    \frac{\partial}{\partial t}(h v_1) + \frac{\partial}{\partial x}\left(h v_1^2 + \frac{g}{2}h^2\right)
+    + \frac{\partial}{\partial y}(h v_1 v_2) + g h \frac{\partial b}{\partial x} &= 0 \\
+    \frac{\partial}{\partial t}(h v_2) + \frac{\partial}{\partial x}(h v_1 v_2)
+    + \frac{\partial}{\partial y}\left(h v_2^2 + \frac{g}{2}h^2\right) + g h \frac{\partial b}{\partial y} &= 0.
+\end{aligned}
+```
+The unknown quantities of the SWE are the water height ``h`` and the velocities ``\mathbf{v} = (v_1, v_2)^T``.
+The gravitational constant is denoted by `g` and the (possibly) variable bottom topography function ``b(x,y)``.
+Conservative variable water height ``h`` is measured from the bottom topography ``b``, therefore one
+also defines the total water height as ``H = h + b``.
+
+The bottom topography function ``b(x,y)`` is set inside the initial condition routine
+for a particular problem setup. To test the conservative form of the SWE one can set the bottom topography
+variable `b` to zero.
+
+In addition to the unknowns, Trixi currently stores the bottom topography values at the approximation points
+despite being fixed in time. This is done for convenience of computing the bottom topography gradients
+on the fly during the approximation as well as computing auxiliary quantities like the total water height ``H``
+or the entropy variables.
+This affects the implementation and use of these equations in various ways:
+* The flux values corresponding to the bottom topography must be zero.
+* The bottom topography values must be included when defining initial conditions, boundary conditions or
+  source terms.
+* [`AnalysisCallback`](@ref) analyzes this variable.
+* Trixi's visualization tools will visualize the bottom topography by default.
+
+References for the SWE are many but a good introduction is available in Chapter 13 of the book:
+- Randall J. LeVeque (2002)
+  Finite Volume Methods for Hyperbolic Problems
+  [DOI: 10.1017/CBO9780511791253](https://doi.org/10.1017/CBO9780511791253)
+"""
+struct ShallowWaterEquations2D{RealT<:Real} <: AbstractShallowWaterEquations{2, 4}
+  gravity::RealT # gravitational constant
+end
+
+function ShallowWaterEquations2D(gravity_constant)
+  ShallowWaterEquations2D(gravity_constant)
+end
```

Comment:
Huh - why is this necessary? And why does it even work? I think there should at least be a comment explaining why this is needed...

Reply:
It is useful to be able to set the gravity constant depending on the application. Most use either `1.0` or `9.81` but some "standard" tests use the value `9.812` while others use `9.8` or `10.0`. I can add a comment about this.

Reply:
I guess @sloede was thinking about something like
```julia
julia> struct ShallowWaterEquations2D{T}
           gravity_constant::T
       end

julia> function ShallowWaterEquations2D(gravity_constant)
         ShallowWaterEquations2D(gravity_constant)
       end
ShallowWaterEquations2D

julia> ShallowWaterEquations2D(1.0)
ERROR: StackOverflowError:
Stacktrace:
 [1] ShallowWaterEquations2D(gravity_constant::Float64) (repeats 79984 times)
   @ Main ./REPL[2]:2
```
It's clear that we need to be able to set some `gravity_constant`.

Reply:
Ah, so his comment was more aimed at "why is this working"? Should I adapt this to look more like the compressible Euler setup and how it sets `gamma`? The current shallow water strategy is sort of a hold over from my original way of handling the bottom topography by passing a function.

Reply:
I guess the basic question is: Do we need this outer constructor at all?

Reply:
> I guess the basic question is: Do we need this outer constructor at all?

Sorry for the confusion about my comment. @ranocha's comment is exactly what I was wondering about. If anything, I can imagine an outer convenience constructor such as
```suggestion
ShallowWaterEquations2D(; gravity_constant) = ShallowWaterEquations2D(gravity_constant)
```
to allow changing the gravity constant in an elixir using `trixi_include`. But is this even useful? If it is a quasi-constant anyways, without practical reasons to change it, you could even consider to set `gravitiy_constant=9.81`) as the default...

Reply:
I kept the constructor external and added another field to it. I use this quantity to test well-balancedness error